### ISSUE TITLE
Research Gist API

### ISSUE THREAD

**Issue by curran** [datavis-tech/vizhub-issue-tracker#301]
* How to create new Gist?
 * How to write/commit to a Gist?
 * How to get latest content from a Gist?

**Comment by denisemauldin** [comment]
The Gist API documentation is located here:

https://docs.github.com/en/rest/reference/gists

To create a gist the user must be logged in to GitHub and the user must grant the `gist` OAuth scope.  
It looks like the scope vizhub requests currently is for 'public data only', so it would need to be updated to add the `gist` scope.

If the `truncated` key is True in any of the response objects it means that either the full list of files was not provided or the full file contents were not provided.

Creating a gist is a POST to /gists of type `application/vnd.github.v3+json` with a `files` object that contains a filename as a key (string) and the contents of the file as the value.  Optional keys are `description` containing a string that serves as a name for the gist and `public` which is a boolean to determine if the gist shows up in the github search or not.  Gists are not truly private because if you have the link for them they can be seen by anyone.

A users's gists can be listed by doing a GET to /gists for the authenticated user.

The contents of a gist can be listed by doing a GET to /gists/<gist_id> 
A gist can be updated by doing a PATCH to /gists/<gist_id> with the same information as required in the create.
A gist can be deleted by doing a DELETE to /gists/<gist_id>

There are a variety of libraries that support writing to gists/github repositories, but the most full featured seems to be using the Passport.js library and the passport-github strategy.
http://www.passportjs.org/

https://github.com/jaredhanson/passport
https://github.com/jaredhanson/passport-github

The advantage of Passport is that it allows you to auth via many different kinds of strategies using OAuth or OpenID.

So we could choose to roll a simple implementation that does POST/GET/PATCH/DELETE to the GitHub Gist API or we could implement an authentication library like Passport.  

Any further questions?  Which direction would you like to go?  Should I spend any additional time mocking up either option?

**Comment by curran** [comment]
Nice sleuthing @denisemauldin !

We do have a `description` field already, which can be used in the Gist `description` field.

I believe Gists also have a `title` field, which we can also populate from our `title` field.

Re: Passport, I don't think it really gives us anything. Sure, there is an option for Gist-write permissions in the API of `passport-github`, but that library does not contain the functionality that actually writes or reads from Gists (it's purely authentication AFIAK). I recall deciding against Passport years ago and opting for a solution that uses JWT (Json Web Tokens).

Also, I'm not sure if we want to enable the Gist write permissions for everyone that logs into VizHub straight away. We could, and that would be the easiest thing to do. However ideally we'd auth as we do now, and only when a user wants to write to a Gist would we ask them to grant the additional permissions. The same goes for GitHub repository integration - we would not want to ask the user to grant permissions to access their GitHub repositories unless they want to explicitly enable that feature.

So I'd say please go in the following direction:
 * No Passport
 * Only upgrade to Gist read & write permissions when the user tries to export to a Gist
 * Go for a first pass implementation of a bare bones demo that writes the content of a Viz to a Gist
 * Once a first pass proof-of-concept is working, we can iterate from there

Sounds good? Thanks!

**Comment by denisemauldin** [comment]
@curran Ok.  What's the expected user flow when someone who has not given `gist` permissions tries to export to Github?  It goes through another OAuth popup flow that's similar to `signInFlow` that uses the `AuthPage`?  So upgrading that flow to have an optional `scope` parameter?

**Comment by curran** [comment]
@denisemauldin Yes, that sounds right to me. Let's get that working and see how it feels.

It may be jarring to the user to be directed immediately to the auth flow, so maybe in the case when someone who has not given `gist` permissions tries to export to Github, there could be a modal that pops up explaining the situation to the effect of:

> To export to GitHub, you first need to grant permissions for VizHub to access Gists.

Followed by a button that says "Grant permission", which takes the user into the auth flow.

But this is the sort of thing we can do in phases.

 * Phase I - get the auth popup flow to basically work, as you described.
 * Phase II - add the modal that orients the user to what's goingon.

Or if it's easier in the long run to do these both together, go for it!

In any case, please put this work behind a feature flag, so we can merge & deploy WIP with the flag disabled. This means, just a boolean exported from `packages/neoFrontend/src/featureFlags.js` that enables the new UI elements to show up.

Thanks! Looking forward to see this progress.

**Comment by curran** [comment]
@denisemauldin I did a little poking around in CodeSandbox for UI/UX ideas. Here's what I found:

![image](https://user-images.githubusercontent.com/68416/88678697-bb1eac00-d0bc-11ea-98dd-2d6fc2fa9feb.png)

![image](https://user-images.githubusercontent.com/68416/88678777-c96cc800-d0bc-11ea-9039-a6bf6156e878.png)

![image](https://user-images.githubusercontent.com/68416/88678856-e2757900-d0bc-11ea-81ff-2a439015fd5c.png)

![image](https://user-images.githubusercontent.com/68416/88678942-f7520c80-d0bc-11ea-8bd2-d363e5f9bb78.png)

![image](https://user-images.githubusercontent.com/68416/88679023-0933af80-d0bd-11ea-9b3b-c2d525db25df.png)

![image](https://user-images.githubusercontent.com/68416/88679104-1b155280-d0bd-11ea-8736-de5f26a5c3fa.png)

![image](https://user-images.githubusercontent.com/68416/88679535-865f2480-d0bd-11ea-96ad-e8b102a260c6.png)

![image](https://user-images.githubusercontent.com/68416/88679611-97a83100-d0bd-11ea-8dfe-9a9773b6f794.png)

![image](https://user-images.githubusercontent.com/68416/88679755-c1f9ee80-d0bd-11ea-83d4-324022be234e.png)

**Comment by curran** [comment]
More UI for inspiration:

https://vega.github.io/editor/?#/

![image](https://user-images.githubusercontent.com/68416/83451324-b4f5b180-a424-11ea-911d-afc7f454d6e7.png)

![image](https://user-images.githubusercontent.com/68416/83451359-cccd3580-a424-11ea-9a28-e043647b953e.png)

Some great ideas there.

**Comment by denisemauldin** [comment]
Ok, I added a PR for a WIP for some of the updated flow.  I'm not sure how to update `AuthContext`.  The goal is for there to be able to be a component in the visualization that looks something like this:

![](https://user-images.githubusercontent.com/68416/88679023-0933af80-d0bd-11ea-9b3b-c2d525db25df.png)

and the `Sign In` button calls `SignInFlow` with the `scopes` set to `['repo']` to check for the correct authorization.

**Comment by curran** [comment]
Nice! Love the direction here. Indeed let's go for repo integration first.

**Comment by curran** [comment]
Forum discussion: [New feature- Github integration](https://vizhub.com/forum/t/new-feature-github-integration/19/2)
